docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,386 @@
+change becomes visible to SQL clients.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC. It could be possible to support foreign
+keys in the future by cascading the type change, but this would greatly
+increase the amount of coordination required to effect the total schema
+change.
+
+### Interleaved tables
+
+Changing the types of primary keys is not within the scope of this RFC,
+so this feature does not interact with interleaved tables. When changes
+to primary keys are supported, a change to the PK of a parent table
+would have to be cascaded to the corresponding PK prefix column of any
+interleaved descendant tables.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Views
+
+Columns that are depended upon by a `CREATE VIEW` are out of scope for
+this RFC. A future change to support views should be possible in the
+same manner as supporting columns that are the target of a foreign-key
+constraint.
+
+# Reference-level explanation
+
+## Process
+
+Conceptually, a column type change can be thought of as creating an
+anonymous, "shadow" column of the new type, which captures writes to the
+original, source, column while a backfill process is run. Once the
+backfill is complete, the shadow column is swapped for the original,
+which is then retired. The shadow column will be generally invisible to
+the user until it is swapped in. Shadow columns will be created in the
+same column family as the source column.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER ... COLUMN TYPE` is run, resulting in a new shadow column being
+  created. The shadow column will be created as a computed column, in
+  the source column's family, with the expression provided by the user
+  via `USING expression` or inferred for cast-type conversions.
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
+* The shadow column and indexes are backfilled using existing
+  functionality for creating
+  [new computed columns](https://github.com/cockroachdb/cockroach/issues/22652)
+  and indexes.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column and
+  indexes.
+  * If the new value that is written is incompatible with the requested
+    conversion (e.g. `'foo'` -> `INET`), an error will be returned to
+    the caller.
+* Once the backfill is complete, several actions happen atomically:
+  * The computed nature is removed from the shadow column.
+  * The hidden flag is removed from the shadow column.
+  * The source column is dropped.
+  * The shadow column is renamed.
+  * Any indexes on the original column are dropped.
```

Comment:
This will break any queries that specify the index by name in a query. I think that's ok though, and we can document it. I'm just thinking through this.

Reply:
Indeed. Also indexes have a DependedOnBy populated by views, this needs to be taken into account too.

Reply:
We should be able to rename the shadow indexes while we're mucking around in those data structures, right?

Reply:
Yes indeed. renaming the shadow index(es) to be the original index name(s) should do the trick.